pvlib/horizon.py

```diff
@@ -0,0 +1,649 @@
+        Longitude values at each point on the grid. Values will vary along
+        axis=0 and will be constant along axis=1.
+    '''
+
+    lat_start = lat - (grid_radius * grid_step)
+    lat_stop = lat + (grid_radius * grid_step)
+    lats = np.linspace(lat_start, lat_stop, 2*grid_radius + 1)
+
+    lon_start = lon - (grid_radius * grid_step)
+    lon_stop = lon + (grid_radius * grid_step)
+    lons = np.linspace(lon_start, lon_stop, 2*grid_radius + 1)
+
+    lat_grid, lon_grid = np.meshgrid(lats, lons)
+
+    return lat_grid, lon_grid
+
+
+def elevation_and_azimuth(pt1, pt2):
+    '''
+    Calculates the elevation angle from pt1 to pt2. Elevation angle is
+    defined as the angle between the line connecting pt1 to pt2 and the plane
+    tangent to the Earth's surface at pt1. A point that appears above the
+    horizontal has a positive elevation angle. Also computes the azimuth
+    defined as degrees East of North of the bearing from pt1 to pt2.
+    This uses the Haversine formula.
+
+    Parameters
+    ----------
+    pt1 : ndarray
+        Nx3 array that contains latitude, longitude, and elevation values
+        that correspond to the origin (observer) points from which the
+        elevation angles are to be calculated. Longitude should be given in
+        degrees East of the Prime Meridian and latitude in degrees North of the
+        Equator. Units are [deg, deg, meters]
+
+
+    pt2 : ndarray
+        Nx3 array that contains latitude, longitude, and elevation values
+        that correspond to the target (observee) points to which the elevation
+        angles are to be calculated. Longitude should be given in
+        degrees East of the Prime Meridian and latitude in degrees North of the
+        Equator. Units are [deg, deg, meters]
+
+    Returns
+    -------
+    bearing_deg: numeric
+        The bearings from pt1 to pt2 in degrees East of North.
+
+    elevation_angle_deg: numeric
+        The elevation angles that the points in pt2 make with the horizontal
+        as observed from the points in pt1. Given in degrees above the
+        horizontal.
+
+    '''
+    a = 6378137.0
+    b = 6356752.0
+
+    lat1 = np.atleast_1d(pt1.T[0])
+    lon1 = np.atleast_1d(pt1.T[1])
+    elev1 = np.atleast_1d(pt1.T[2])
+    lat2 = np.atleast_1d(pt2.T[0])
+    lon2 = np.atleast_1d(pt2.T[1])
+    elev2 = np.atleast_1d(pt2.T[2])
+
+    # convert to radians
+    phi1 = np.radians(lat1)
+    theta1 = np.radians(lon1)
+    phi2 = np.radians(lat2)
+    theta2 = np.radians(lon2)
+
+    v1 = tools.lle_to_xyz(np.stack([lat1, lon1, elev1], axis=1))
+    v2 = tools.lle_to_xyz(np.stack([lat2, lon2, elev2], axis=1))
+    x1 = np.atleast_1d(v1.T[0])
+    y1 = np.atleast_1d(v1.T[1])
+    z1 = np.atleast_1d(v1.T[2])
+
+    delta = np.atleast_2d(np.subtract(v1, v2))
+
+    normal = np.atleast_2d(np.stack([2*x1/a**2, 2*y1/a**2, 2*z1/b**2], axis=1))
+
+    # Take the dot product of corresponding vectors
+    dot = np.sum(np.multiply(delta, normal), axis=1)
```

Comment:
Would `np.dot()` do the trick?

Reply:
Okay so I'm taking two Nx3 arrays and trying to do an "element-wise" dot product to get back an Nx1 array. I couldn't get `np.dot()` to do this but let me know if you know how.